### ISSUE TITLE
关于mini‐batched策略，请询问一下

### ISSUE THREAD

**Issue by chenzhaoxingxin** [liufeng2317/TorchInversion#12]
作者您好：
  很荣幸看到您在JGR上的文章，我看到文章中提到使用了“mini‐batch processing”，然后在文章中提到了“Note we first accumulate the gradients  from all individual mini‐batched sources before updating the model”。
我看到github上分享的程序：

![Image](https://github.com/user-attachments/assets/90e6d4a7-837a-486c-a751-36527573726d)
这个位置，应该就是将总的震源分批求解梯度吧，同时下面的代码应该就是利用backward求解梯度：

![Image](https://github.com/user-attachments/assets/6d385638-a78e-4605-9285-2d02e2975041)
我想请教一下，“累积所有单个小批量源的梯度”是哪一步实现的呢？
希望不吝赐教！

**Comment by liufeng2317** [comment]
你好！

非常感谢你的提问。你提供的第一张图展示了将所有炮记录划分为不同的批次（batch），并分别进行波场正演与梯度计算的过程。第二张图则是对所有批次计算得到的梯度进行后处理，并不代表梯度的求解过程本身。关于梯度的累积，这里采用的是典型的梯度累积训练策略。每一个 batch 计算一次 `loss`，并调用 `loss.backward()` 计算该 batch 的梯度。只要在调用 `optimizer.step()` 之前不执行 `optimizer.zero_grad()`，这些梯度就会在多个 batch 之间自动累积。直到处理完所有 batch 后，调用 `optimizer.step()` 利用累计的梯度来更新模型参数。

![Image](https://github.com/user-attachments/assets/1828bfeb-786b-4b93-870f-7564bf008b8d)

**Comment by liufeng2317** [comment]
github中我建议采用 checkpointing 策略，其主要优势在于：通过灵活设置检查点数量，可以显著降低内存占用，而对计算效率的影响却相对较小。在同等条件下，尽管 mini-batch 策略 也能线性降低内存开销，但它通常会带来线性增长的计算时间。因此，相较而言，checkpointing 更适合在资源受限的场景下实现效率与内存之间的折中。

**Comment by chenzhaoxingxin** [comment]
感谢回复！我再仔细阅读文章，深入了解一下checkpointing 策略，以前了解该策略是关于有限差分重建逆时波场的，不知道是否类似？

该位置处，每次反演迭代刚开始处使用self.optimizer.zero_grad()， 是不是起清空梯度作用的？
![Image](https://github.com/user-attachments/assets/e962d71c-b284-49e9-933a-b551ff95a784)

**Comment by liufeng2317** [comment]
是的，每次迭代需要清空梯度，以避免每次迭代的梯度都累积占用过多内存（等价于正常深度神经网络的训练过程）。关于checkpoints策略可以参考：
文章：Gruslys, A., Munos, R., Danihelka, I., Lanctot, M., & Graves, A. (2016). Memory-efficient backpropagation through time. In Proceedings of the 30th international conference on neural information processing systems (pp. 4132–4140). Curran Associates Inc.
代码：https://pytorch.org/docs/stable/checkpoint.html

**Comment by liufeng2317** [comment]
是的，每次迭代需要清空梯度，以避免梯度累积占用过多内存（等价于深度神经网络的训练过程的迭代）。关于checkpoints策略可以参考：
文章：Gruslys, A., Munos, R., Danihelka, I., Lanctot, M., & Graves, A. (2016). Memory-efficient backpropagation through time. In Proceedings of the 30th international conference on neural information processing systems (pp. 4132–4140). Curran Associates Inc.
代码：https://pytorch.org/docs/stable/checkpoint.html
